config-common/src/watcher.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use async_trait::async_trait;
+use notify::Watcher;
+use std::path::PathBuf;
+
+type Api<TConfig, TError> = std::sync::Arc<
+    futures_util::lock::Mutex<dyn UpdateConfig<Config = TConfig, Error = TError> + Send>,
```

Comment:
Why does this need to be a trait object? The user of this, `start_watcher` doesn't need it to be one. `start_watcher` is already generic anyway, on `TConfig` and `TError`.

```rust
fn start_watcher<TApi>(...)
where
    TApi: UpdateConfig,
    <TApi as UpdateConfig>::Config: serde::de::DeserializeOwned + Send,
    <TApi as UpdateConfig>::Error: std::error::Error,
{
```

Reply:
Looks like `UpdateConfig`'s trait definition already includes the `serde::de::DeserializeOwned` and `std::error::Error` bounds, so I'm pretty sure you won't need those two extra bounds on `start_watcher`.

Oh, but `UpdateConfig::::Config` doesn't have a `Send` bound in the trait definition...
IMO it's probably fine to add a `Send` bound to the trait definition in order to simplify `start_watcher`'s trait bounds. After all, I can't imagine a situation where TConfig _wouldn't_ be `Send`.

Reply:
Fixed

Reply:
It hasn't been fixed? All I see is you've inlined the typedef into the function definition. It's still a trait object.

Reply:
Sorry, misread what you wrote the first time.  Fixed.